climada/util/coordinates.py

```diff
@@ -1996,105 +2024,165 @@ def read_raster_bounds(path, bounds, res=None, bands=None, resampling="nearest",
+        xres = ONE_LAT_KM * 1000 * xres * np.cos(np.radians(lat))
+        yres = ONE_LAT_KM * 1000 * yres
+
+    diff_x = np.diff(data, axis=1)
+    diff_x = 0.5 * (diff_x[1:, :] + diff_x[:-1, :])
+    diff_y = np.diff(data, axis=0)
+    diff_y = 0.5 * (diff_y[:, 1:] + diff_y[:, :-1])
+    gradient_data = np.stack([diff_y / yres, diff_x / xres], axis=-1)
+
+    return gradient_data, gradient_transform
+
+def read_raster_sample(path, lat, lon, intermediate_res=None, method=('linear', 'nearest'),
+                       fill_value=None, gradient=False):
     """Read point samples from raster file.
 
+    To avoid redundant IO operations, the gradients at the specified points can be computed in the
+    same function call using the `gradient` feature. For example, in case of an elevation data set,
+    the slopes of the terrain in x- and y-direction are returned. In addition, if the CRS of the
+    elevation data set is EPSG:4326 (lat/lon) and elevations are given in m, then distances are
+    converted from degrees to meters, so that the unit of the returned slopes is "meters (height)
+    per meter (distance)".
+
     Parameters
     ----------
     path : str
         path of the raster file
-    lat : np.array
+    lat : np.array of shape (npoints,)
         latitudes in file's CRS
-    lon : np.array
+    lon : np.array of shape (npoints,)
         longitudes in file's CRS
-    intermediate_res : float, optional
+    intermediate_res : float or pair of floats, optional
         If given, the raster is not read in its original resolution but in the given one. This can
         increase performance for files of very high resolution.
-    method : str, optional
-        The interpolation method, passed to scipy.interp.interpn. Default: 'linear'.
+    method : str or pair of str, optional
+        The interpolation methods for the data and its gradient, passed to
+        `scipy.interpolate.interpn`. If a single string is given, the same interpolation method is
+        used for both the data and its gradient. Default: ('linear', 'nearest')
     fill_value : numeric, optional
         The value used outside of the raster bounds. Default: The raster's nodata value or 0.
+    gradient : boolean, optional
+        If True, compute the raster gradient at the specified points. For convenience, and because
+        this is the most common use case, the step sizes in the gradient computation are converted
+        to meters if the raster's CRS is EPSG:4326 (lat/lon). Default: False
 
     Returns
     -------
-    values : np.array of same length as lat
+    values : np.array of shape (npoints,)
         Interpolated raster values for each given coordinate point.
+    gradient : np.array of shape (npoints, 2), optional
+        If grad=True, the raster gradient at each of the given coordinate points is returned.
+        The first/second value in each row is the derivative in lat/lon direction (lat is first!).
```

Comment:
I prefer methods returning unambiguous data types, or if need be an Option thereof. So that after `y = f(x)` `y` is always of type `T` (or `None`). Here `y` could be an array or a pair of arrays.

In this particular case I would have preferred two methods sharing the job, something like this:

```python
values = read_raster_sample_without_gradient(...)
gradient = read_raster_sample_gradient(values, ...)
```

Since I assume that would be a waste of cpu in some way, I'd suggest to always return a pair 
```python 
def read_raster_sample(...) -> Tuple[np.array, Option[np.array]]:
    ...
```

Reply:
I solved this problem by leaving the old functionality of `read_raster_sample` untouched, and adding a new function `read_raster_sample_with_gradients`. There is almost no redundant code, because I moved some of the code to a helper function that is used by both functions.

Reply:
👍 _much_ better than my suggestion as it keeps backward compatibility!